### PR TITLE
raft: reduce memory allocation when reset

### PR DESCRIPTION
**Pull request by BusyJay**
Create a new progress will allocate a vec. So this pr adds a reset function to avoid the allocation. Little performance enhancement is detected:

```
Raft::campaign (7, 3, CampaignElection)
                        time:   [2.8076 us 2.8291 us 2.8565 us]
                        change: [-4.6982% -4.2007% -3.6530%] (p = 0.00 < 0.05)
                        Performance has improved.
```

### CODE REVIEW

**Comment by Hoverbear** [comment]
Looks like this depends on #119 ?

File: src/progress.rs [hunk 1]
```diff
@@ -233,7 +233,7 @@ impl ProgressSet {
 }
 
 /// The progress of catching up from a restart.
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Clone)]
```
Comment (Hoverbear) [inline src/progress.rs]: Does a `Progress` have a sane defautt?

File: src/progress.rs [hunk 2]
```diff
@@ -402,7 +417,7 @@ impl Progress {
 }
 
 /// A buffer of inflight messages.
-#[derive(Debug, Default, Clone, PartialEq)]
+#[derive(Debug, Clone, PartialEq)]
```
Comment (Hoverbear) [inline src/progress.rs]: Similarly, is it possible to make a sane default?

File: src/progress.rs (hunk 1 above)
Comment (BusyJay) [reply src/progress.rs]: Yes. An explicit capacity needs to be specified to avoid easily wrong configuration.

Comment (Hoverbear) [reply src/progress.rs]: Ok :+1:

**Comment by siddontang** [comment]
PTAL @hicqu

**Comment by hicqu** [comment]
LGTM. 👍

**Comment by siddontang** [comment]
@Hoverbear please approves it

**Pull request merged**